### PR TITLE
Bug 1866318 - Retry intent referrer access once

### PR DESCRIPTION
**Pull request by MorMundHS-MA**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1866318

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/IntentReceiverActivity.kt [hunk 1]
```diff
@@ -140,6 +142,25 @@ class IntentReceiverActivity : Activity() {
...
+        } catch (e: RuntimeException) {
+            try {
+                referrer
+            } catch (e: Exception) {
+                null
+            }
+        } catch(e: Exception) {
+            null
+        }
+    }
```
Comment (MorMundHS-MA) [inline fenix/app/src/main/java/org/mozilla/fenix/IntentReceiverActivity.kt:162]: For some reason, the second access does not cause a `RuntimeException` for the `referrer` access. It does not seem to be timing based, as waiting with a break point does not prevent the error from happening. Inspecting the variable does however throw the first time as well, but not the second time. 

However this whole behavior might be very specific to Spotify's implementation and I'm not sure how important the referrer information is. It seems to work fine without it, so a less complex fix is a simple try catch around the `referrer` access.

**Comment by jonalmeida** [comment]
Hi @MorMundHS-MA , thanks for the patch! But as you've noticed, this patch might not be the best solution.

We have had situations like this before with all intents that can come from third-party apps. We've found it's safer to `SafeIntent` which catches these kind of exceptions. Could you add another accessor for the `referrer`.

It might be a bit tricky because I can see there are some extension functions like `sanitize` and `stripUnwantedFlags` on the [`IntentReceiverActivity`](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/IntentReceiverActivity.kt#47-48).

Hopefully, this is a helpful direction to getting this patch ready for landing. Please feel free to ping me if I can help more. 🙂

**Review by jonalmeida — DISMISSED** [review]
Hi @MorMundHS-MA , thanks for the patch! But as you've noticed, this patch might not be the best solution.

We have had situations like this before with all intents that can come from third-party apps. We've found it's safer to `SafeIntent` which catches these kind of exceptions. Could you add another accessor for the `referrer`.

It might be a bit tricky because I can see there are some extension functions like `sanitize` and `stripUnwantedFlags` on the [`IntentReceiverActivity`](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/IntentReceiverActivity.kt#47-48). (EDIT: looks like [focus converts to a SafeIntent after santizing](https://searchfox.org/mozilla-mobile/source/firefox-android/focus-android/app/src/main/java/org/mozilla/focus/activity/IntentReceiverActivity.kt#28))

Hopefully, this is a helpful direction to getting this patch ready for landing. Please feel free to ping me if I can help more. 🙂

**Comment by MorMundHS-MA** [comment]
@jonalmeida Hi, thanks for the information! It gave me a direction to look further into the issue.

Unfortunately using `SafeIntent` won't really help imo. `getReferrer` is a method of the `Activity` itself not the intent. It *does* potentially get its information from the `Intent` ([AOSP code](https://github.com/aosp-mirror/platform_frameworks_base/blob/fe9c3c82cec4c0b7149fa40cd5d3934d172b1177/core/java/android/app/Activity.java#L6768)). But from my testing the actual value of `getReferrer` comes from the `mReferrer` variable in the `Activity`.

So I could use the `SafeIntent` to copy most of the `getReferrer` behavior, this is [what chromium does](https://github.com/chromium/chromium/blob/96df497cc1081929f79c544c5e89df3911da86b9/chrome/android/java/src/org/chromium/chrome/browser/IntentHandler.java#L480). This has the benefit of working on all SDK versions, where as `getReferrer` needs API>=22. But it does not get the referrer in most cases (which I'm still not sure how important the information really is).

Alternatively, I could add a `getSafeReferrer` extension method to `Activity` to mirror the design of `toSafeIntent` somewhat. That would still leave the question if I should fix the Spotify bug where it throws only the first time the `Intent` is accessed or if I should ignore it.

Lastly, there is always the option of reading `mReferrer` through reflection, which I don't think anyone would consider a good idea.

Overall I think my original solution is still the overall cleanest, unfortunately.

**Comment by jonalmeida** [comment]
I don't have capacity to review this so leaving it up to the Fenix team to take over.

**Comment by gabrielluong** [comment]
Hello @MorMundHS-MA , Thank you so much for your contribution. We have moved our firefox-android repo over to [mozilla-central](https://hg.mozilla.org/mozilla-central). Please feel free to reopen a new PR against the new repo. You will have to clone the new repository and reapply your changes. Thank you again, and sorry for the inconvenience.

For more information, please see https://github.com/mozilla-mobile/firefox-android/wiki#upcoming-migration-to-mozilla-central.

**Pull request closed without merging**